File: zig/3-pid-bm/main/perf.zig
const std = @import("std");
const idf = @import("esp_idf");
const sys = idf.sys;

const c = @import("c.zig");
const utils = @import("utils.zig");

pub const StartMarker = struct {
    cycle: sys.esp_cpu_cycle_count_t,

    pub fn now() StartMarker {
        return .{ .cycle = sys.esp_cpu_get_cycle_count() };
    }
};

pub const Counter = struct {
    const Self = @This();

    name: []const u8,
    cpu_frequency: u32,
    total_cycles: sys.esp_cpu_cycle_count_t,
    sample_count: u32,

    pub fn init(name: []const u8) !Self {
        var cpu_frequency: u32 = undefined;
        try c.espCheckError(c.esp_clk_tree_src_get_freq_hz(
            c.SOC_MOD_CLK_CPU,
            0,
            &cpu_frequency,
        ));

        return .{
            .name = name,
            .cpu_frequency = cpu_frequency,
            .total_cycles = 0,
            .sample_count = 0,
        };
    }

    pub fn add_sample(self: *Self, start: StartMarker) void {
        const end = sys.esp_cpu_get_cycle_count();
        const cycles = end - start.cycle;

        self.total_cycles += cycles;
        self.sample_count += 1;
    }

    pub fn report(self: *const Self) void {
        const cycles_avg = @as(f64, @floatFromInt(self.total_cycles)) / @as(f64, @floatFromInt(self.sample_count));
        const time_ms = cycles_avg / @as(f64, @floatFromInt(self.cpu_frequency)) * 1e6;
        std.log.info(
            "Performance counter {s}: {d:.3} us = {d:.0} cycles ({} sampl.)",
            .{ self.name, time_ms, cycles_avg, self.sample_count },
        );
    }

    pub fn reset(self: *Self) void {
        self.total_cycles = 0;
        self.sample_count = 0;
    }
};

File: rust/3-pid-bm/src/perf.rs
use core::arch::asm;
use esp_idf_sys::{
    esp, esp_clk_tree_src_get_freq_hz, esp_cpu_cycle_count_t, soc_module_clk_t_SOC_MOD_CLK_CPU,
};
use log::info;
use std::mem::MaybeUninit;

pub struct Measurement<'a> {
    counter: &'a mut Counter,
    start: esp_cpu_cycle_count_t,
}

impl Drop for Measurement<'_> {
    fn drop(&mut self) {
        self.counter.add_sample(self.start);
    }
}

pub struct Counter {
    name: &'static str,
    cpu_frequency: u32,
    total_cycles: esp_cpu_cycle_count_t,
    sample_count: u32,
}

impl Counter {
    pub fn new(name: &'static str) -> anyhow::Result<Self> {
        let mut cpu_frequency = MaybeUninit::uninit();
        esp!(unsafe {
            esp_clk_tree_src_get_freq_hz(
                soc_module_clk_t_SOC_MOD_CLK_CPU,
                0,
                cpu_frequency.as_mut_ptr(),
            )
        })?;

        Ok(Counter {
            name,
            cpu_frequency: unsafe { cpu_frequency.assume_init() },
            total_cycles: 0,
            sample_count: 0,
        })
    }

    pub fn measure(&mut self) -> Measurement {
        Measurement {
            counter: self,
            start: esp_cpu_get_cycle_count(),
        }
    }

    fn add_sample(&mut self, start: esp_cpu_cycle_count_t) {
        let end = esp_cpu_get_cycle_count();
        let cycles = end - start;

        self.total_cycles += cycles;
        self.sample_count += 1;
    }

    pub fn report(&self) {
        let cycles_avg = self.total_cycles as f64 / self.sample_count as f64;
        let time_us = cycles_avg / self.cpu_frequency as f64 * 1e6;

        info!(
            "Performance counter {}: {:.3} us = {:.0} cycles ({} sampl.)",
            self.name, time_us, cycles_avg, self.sample_count
        );
    }

    pub fn reset(&mut self) {
        self.total_cycles = 0;
        self.sample_count = 0;
    }
}

macro_rules! rsr {
    ($reg:expr) => {{
        let result: u32;
        asm! {
            "rsr {ret}, {reg}",
            ret = out(reg) result,
            reg = const $reg,
        }
        result
    }};
}

#[inline]
fn esp_cpu_get_cycle_count() -> u32 {
    unsafe { rsr!(234) }
}
